`possibly-used-before-assignment`  does not handle `match` in nested `if`

### Bug description

```python
# pylint_used_before_assignment_nested_match.py

from enum import Enum
from typing import assert_never


class Example(Enum):
    FOO = 1
    BAR = 2


def check_value_only_match(example: Example) -> str | None:
    match example:
        case Example.FOO:
            result = "foo"
        case Example.BAR:
            result = "bar"
        case _:
            assert_never(example)

    return result


def check_value_if_then_if(example: Example, should_check: bool) -> str | None:
    if should_check:
        result = None
    else:
        if example == Example.FOO:
            result = "foo"
        elif example == Example.BAR:
            result = "bar"
        else:
            assert_never(example)

    return result


def check_value_if_then_match_return(example: Example, should_check: bool) -> str | None:
    if should_check:
        result = None
    else:
        match example:
            case Example.FOO:
                result = "foo"
            case Example.BAR:
                result = "bar"
            case _:
                return None

    return result


def check_value_if_then_match_raise(example: Example, should_check: bool) -> str | None:
    if should_check:
        result = None
    else:
        match example:
            case Example.FOO:
                result = "foo"
            case Example.BAR:
                result = "bar"
            case _:
                raise ValueError("Not a valid enum")

    return result


def check_value_if_then_match_assert_never(
    example: Example, should_check: bool
) -> str | None:
    if should_check:
        result = None
    else:
        match example:
            case Example.FOO:
                result = "foo"
            case Example.BAR:
                result = "bar"
            case _:
                assert_never(example)

    return result
```


### Configuration

_No response_

### Command used

```shell
% ./venv/bin/pylint --disable=all --enable=possibly-used-before-assignment pylint_used_before_assignment_nested_match.py
```


### Pylint output

```python
pylint_used_before_assignment_nested_match.py:50:11: E0606: Possibly using variable 'result' before assignment (possibly-used-before-assignment)
pylint_used_before_assignment_nested_match.py:65:11: E0606: Possibly using variable 'result' before assignment (possibly-used-before-assignment)
pylint_used_before_assignment_nested_match.py:82:11: E0606: Possibly using variable 'result' before assignment (possibly-used-before-assignment)

------------------------------------------------------------------
Your code has been rated at 6.59/10 (previous run: 7.22/10, -0.63)
```


### Expected behavior

The first two functions are correctly handled by pylint, i.e.
* a regular `match` works
* two nested `if`s work

I though this was a variant of https://github.com/pylint-dev/pylint/issues/9643 at first, with `match` not properly supported, but it's apparently a bit more complicated: it seems pylint has trouble understanding that the `match` nested under the first `if` is also "exhaustive". Note that this is triggered whatever "exit" type is used in the `case _:` fallthrough (`return`, `raise` or `assert_never`)

### Pylint version

```shell
pylint 3.2.2
astroid 3.2.2
Python 3.11.2 (main, Feb 28 2023, 10:36:52) [GCC 12.2.0]
```


### OS / Environment

```
% lsb_release -a
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux trixie/sid
Release:        n/a
Codename:       trixie
```

### Additional dependencies

_No response_

I ran into the same issue after upgrading to pylint: in pylint 3.1.x the problem doesn't occur and since 3.2.0 it does.

Here is a more minimal test case:
```py
def f(x):
    if x is None:
        y = 0
    else:
        match x:
            case int():
                y = x
            case _:
                raise TypeError(type(x))
    return y
```
Output:
```
testcase.py:10:11: E0606: Possibly using variable 'y' before assignment (possibly-used-before-assignment)
```
Versions used:
```
pylint 3.2.3
astroid 3.2.2
Python 3.11.9
```

--- a/doc/whatsnew/3/3.3/index.rst
+++ b/doc/whatsnew/3/3.3/index.rst
@@ -14,6 +14,53 @@ Summary -- Release highlights
 
 .. towncrier release notes start
 
+What's new in Pylint 3.3.8?
+---------------------------
+Release date: 2025-08-09
+
+This patch release includes an exceptional fix for a false negative issue.
+For details, see: https://github.com/pylint-dev/pylint/pull/10482#issuecomment-3164514082
+
+False Positives Fixed
+---------------------
+
+- Fix false positives for `possibly-used-before-assignment` when variables are exhaustively
+  assigned within a `match` block.
+
+  Closes #9668 (`#9668 <https://github.com/pylint-dev/pylint/issues/9668>`_)
+
+- Fix false positive for `missing-raises-doc` and `missing-yield-doc` when the method length is less than docstring-min-length.
+
+  Refs #10104 (`#10104 <https://github.com/pylint-dev/pylint/issues/10104>`_)
+
+- Fix a false positive for ``unused-variable`` when multiple except handlers bind the same name under a try block.
+
+  Closes #10426 (`#10426 <https://github.com/pylint-dev/pylint/issues/10426>`_)
+
+
+
+False Negatives Fixed
+---------------------
+
+- Fix false-negative for ``used-before-assignment`` with ``from __future__ import annotations`` in function definitions.
+
+  Refs #10482 (`#10482 <https://github.com/pylint-dev/pylint/issues/10482>`_)
+
+
+
+Other Bug Fixes
+---------------
+
+- Fix a bug in Pyreverse where aggregations and associations were included in diagrams regardless of the selected --filter-mode (such as PUB_ONLY, ALL, etc.).
+
+  Closes #10373 (`#10373 <https://github.com/pylint-dev/pylint/issues/10373>`_)
+
+- Fix double underscores erroneously rendering as bold in pyreverse's Mermaid output.
+
+  Closes #10402 (`#10402 <https://github.com/pylint-dev/pylint/issues/10402>`_)
+
+
+
 What's new in Pylint 3.3.7?
 ---------------------------
 Release date: 2025-05-04

--- a/doc/whatsnew/fragments/10104.false_positive
+++ b/doc/whatsnew/fragments/10104.false_positive
@@ -1,3 +1,0 @@
-Fix false positive for `missing-raises-doc` and `missing-yield-doc` when the method length is less than docstring-min-length.
-
-Refs #10104

--- a/doc/whatsnew/fragments/10373.bugfix
+++ b/doc/whatsnew/fragments/10373.bugfix
@@ -1,3 +1,0 @@
-Fix a bug in Pyreverse where aggregations and associations were included in diagrams regardless of the selected --filter-mode (such as PUB_ONLY, ALL, etc.).
-
-Closes #10373

--- a/doc/whatsnew/fragments/10402.bugfix
+++ b/doc/whatsnew/fragments/10402.bugfix
@@ -1,3 +1,0 @@
-Fix double underscores erroneously rendering as bold in pyreverse's Mermaid output.
-
-Closes #10402

--- a/doc/whatsnew/fragments/10426.false_positive
+++ b/doc/whatsnew/fragments/10426.false_positive
@@ -1,3 +1,0 @@
-Fix a false positive for ``unused-variable`` when multiple except handlers bind the same name under a try block.
-
-Closes #10426

--- a/doc/whatsnew/fragments/10482.false_negative
+++ b/doc/whatsnew/fragments/10482.false_negative
@@ -1,3 +1,0 @@
-Fix false-negative for ``used-before-assignment`` with ``from __future__ import annotations`` in function definitions.
-
-Refs #10482

--- a/doc/whatsnew/fragments/9668.false_positive
+++ b/doc/whatsnew/fragments/9668.false_positive
@@ -1,4 +1,0 @@
-Fix false positives for `possibly-used-before-assignment` when variables are exhaustively
-assigned within a `match` block.
-
-Closes #9668

--- a/pylint/__pkginfo__.py
+++ b/pylint/__pkginfo__.py
@@ -9,7 +9,7 @@ It's updated via tbump, do not modify.
 
 from __future__ import annotations
 
-__version__ = "3.3.7"
+__version__ = "3.3.8"
 
 
 def get_numversion_from_version(v: str) -> tuple[int, int, int]:

--- a/script/.contributors_aliases.json
+++ b/script/.contributors_aliases.json
@@ -273,6 +273,7 @@
     "mails": [
       "66853113+pre-commit-ci[bot]@users.noreply.github.com",
       "49699333+dependabot[bot]@users.noreply.github.com",
+      "212256041+pylint-backport-bot[bot]@users.noreply.github.com",
       "41898282+github-actions[bot]@users.noreply.github.com"
     ],
     "name": "bot"

--- a/tbump.toml
+++ b/tbump.toml
@@ -1,7 +1,7 @@
 github_url = "https://github.com/pylint-dev/pylint"
 
 [version]
-current = "3.3.7"
+current = "3.3.8"
 regex = '''
 ^(?P<major>0|[1-9]\d*)
 \.

--- a/towncrier.toml
+++ b/towncrier.toml
@@ -1,5 +1,5 @@
 [tool.towncrier]
-version = "3.3.7"
+version = "3.3.8"
 directory = "doc/whatsnew/fragments"
 filename = "doc/whatsnew/3/3.3/index.rst"
 template = "doc/whatsnew/fragments/_template.rst"

